src/main/java/com/thealgorithms/maths/MillerRabinPrimalityCheck.java

```diff
@@ -0,0 +1,118 @@
+package com.thealgorithms.maths;
+
+import java.util.Random;
+
+public class MillerRabinPrimalityCheck {
+
+    /**
+     * Check whether the given number is prime or not
+     * MillerRabin algorithm is probabilistic. There is also an altered version which is deterministic.
+     * https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test
+     * https://cp-algorithms.com/algebra/primality_tests.html
+     *
+     * @param n Whole number which is tested on primality
+     * @param k Number of iterations
+     *       If n is composite then running k iterations of the Miller–Rabin
+     *       test will declare n probably prime with a probability at most 4^(−k)
+     * @return true or false whether the given number is probably prime or not
+     */
+
+    public static boolean millerRabin(long n, int k) { // returns true if n is probably prime, else returns false.
+        if (n < 4) return n == 2 || n == 3;
+
+        int s = 0;
+        long d = n - 1;
+        while ((d & 1) == 0) {
+            d >>= 1;
+            s++;
+        }
+        Random rnd = new Random();
+        for (int i = 0; i < k; i++) {
+            long a = 2 + rnd.nextLong(n) % (n - 3);
+            if (checkComposite(n, a, d, s)) return false;
+        }
+        return true;
+    }
+
+    public static boolean deterministicMillerRabin(long n) { // returns true if n is prime, else returns false.
+        if (n < 2) return false;
+
+        int r = 0;
+        long d = n - 1;
+        while ((d & 1) == 0) {
+            d >>= 1;
+            r++;
+        }
+
+        for (int a : new int[] {2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37}) {
```

Comment:
We need to test all primes that are below a certain limit as far as I understand

Reply:
I don't really get the math behind the deterministic version, but here is some basic explanation of it https://cp-algorithms.com/algebra/primality_tests.html#deterministic-version.

But from here it shows that it doesn't need to be all primes

Reply:
From the link you provided:

> It turns out, for testing a 32 bit integer it is only necessary to check the first 4 prime bases: 2, 3, 5 and 7. The smallest composite number that fails this test is  
$3,215,031,751 = 151 \cdot 751 \cdot 28351$ . And for testing 64 bit integer it is enough to check the first 12 prime bases: 2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, and 37.

Interesting, so they've found them empirically!